docs/rfcs/0006-clustering.md

```diff
@@ -0,0 +1,331 @@
+scale elastically with a single configuration file and an eventually consistent
+model.
+
+## Goals
+* Implement a clustered mode that allows the Agent to elastically scale without
+  changing the configuration
+* Enable Flow components to work together and distribute load across a cluster
+* Enable fine-grained scheduling of Flow components within a cluster
+* Provide an easy-to-use replacement for scraping service and hashmod sharding
+* Allow users to understand and debug the status of their cluster
+
+## Non-goals
+* Recreate the scraping service as-is. More specifically:
+  - Use an external store for configuration files
+  - Expose an API for managing configuration
+  - Running multiple configuration files at once.
+* Distribute load by merging multiple configuration files.
+
+## Proposal
+The proposal is based on prior art: https://github.com/grafana/agent/issues/872, https://github.com/grafana/agent/pull/1140
+
+* We will continue with a
+  [gossip-based](https://en.wikipedia.org/wiki/Gossip_protocol) approach using
+Hashicorp’s memberlist for our cluster
+* We will reuse the rfratto/ckit package code
+* We will use HTTP2 for communication between nodes
+* We will use go-discover for bootstrapping the cluster and discovering peers
+* A non-clustered Agent will work similar to a one-node cluster, which in the
+  future will be the default mode of operation
+
+## Implementation
+The feature will be behind a feature flag `--enable-features=clustering`. An
+agent can opt-in to clustering by passing a `--cluster.discover-peers`
+command-line flag with a comma-separated list of peers to connect to. Whenever
+an agent node receives a message about another node joining or leaving the
+cluster, it will propagate the message to its neighbors, and so on, until this
+information has reached all members of the cluster. The gossip memberlist will
+be utilized for the peer list, health checking and distribution of tokens
+between the agent nodes.
+
+All nodes will have access to a shared ckit.Sharder interface implementation
+which will be used to expose methods for each individual agent’s Flow
+controller to determine ownership of resources. As nodes enter and exit the
+cluster, the Sharder (eg. a consistent hashing ring) will redistribute tokens
+among the nodes in the cluster. The eventually consistent cluster state is when
+all nodes are working with the same configuration and have knowledge of each
+one of their peers in the cluster.The Sharder will be used to determine
+ownership of resources by hashing a desired value and checking the peers
+responsible for the corresponding ring token.
+
+When all nodes in the cluster have an up-to-date image of their peers, they
+will be able to independently agree to the ownership of a resource without
+having to communicate with each other, as local hashing will provide the same
+results for all nodes. For example, an agent node will be able to hash the
```

Comment:
> I'm not sure how this can be true in the general case

Could you expand on that? In our eventual consistent state, when all peers have knowledge of one another, their hashing function will provide the same output for the same input target set.

My current view is that we _don't_ need two set of agent clusters. Redundancy can happen by asking the same hashing function to return N owners for a certain token. This would be configured on the component itself as the clustering provides the overarching mechanism only, and similarly to N=1, no explicit communication or beforehand sharding is required.

Reply:
Sorry, I should have clarified this a bit more. What I mean is that "local hashing will provide the same results", but only if the hashing function is the same and if the "N owners" are the same. Which means that if we want to change either of them, we would need to restart the cluster to make sure it's consistent. I'm not sure if this is something worth mentioning explicitly or we can just assume it's obvious.